components/raftstore/src/store/peer.rs

```diff
@@ -3696,16 +3704,72 @@ where
 
         let transfer_leader = get_transfer_leader_cmd(&req).unwrap();
         let peer = transfer_leader.get_peer();
-
-        let transferred = self.pre_transfer_leader(peer);
-
+        let transferred = if self.peer.id == peer.id {
+            true
+        } else {
+            match self.propose_locks_before_transfer_leader(ctx) {
+                Ok(true) => self.propose_normal(ctx, req).is_ok(),
+                Ok(false) => self.pre_transfer_leader(peer),
+                Err(_) => false,
+            }
+        };
         // transfer leader command doesn't need to replicate log and apply, so we
         // return immediately. Note that this command may fail, we can view it just as an advice
         cb.invoke_with_response(make_transfer_leader_response());
 
         transferred
     }
 
+    // Returns whether it needs to propose the transfer leader command instead of
+    // just sending messages.
+    fn propose_locks_before_transfer_leader<T>(
+        &mut self,
+        ctx: &mut PollContext<EK, ER, T>,
+    ) -> Result<bool> {
+        // 1. Disable in-memory pessimistic locks
+        let mut pessimistic_locks = self.txn_ext.pessimistic_locks.write();
+        let locks = if pessimistic_locks.valid {
+            // FIXME: `valid` should be set back to true if the peer remains being leader after
+            // some time (which means leader transfer fails)
+            pessimistic_locks.valid = false;
+            // If there is no lock when setting `valid` to false, we can guarantee that no lock
+            // needs to be migrated to the new leader. In this case, we can just send a message
+            // to trigger leader transfer instead of making a proposal.
+            if pessimistic_locks.map.is_empty() {
+                return Ok(false);
+            }
+            mem::take(&mut pessimistic_locks.map)
+        } else {
+            // If `valid` is false, the locks may have been proposed but not committed.
+            // So, it is not safe to use a Raft message to transfer leader in case the lock
+            // proposal fails.
+            return Ok(true);
+        };
+        drop(pessimistic_locks);
+
+        // 2. Propose pessimistic locks
```

Comment:
Is it possible that the leader fails to transfer leadership due to `leader_transfer_max_log_lag` even if you propose the transfer-leader request other than using raft message? I'm thinking can we propose memory lock requests after the check and add some actions to reduce the unavailable duration due to it.

Reply:
I think there should be no unavailable duration. Do you mean the duration of not being able to write in-memory pessimistic locks?

Reply:
No. The unavailable duration is caused by leadership transfer in which the leader refuses to accept proposals until the transferee catches up with the leader or timeout. If there are many memory-lock requests, it may take a while for the transferee to catch up.

Reply:
The `TransferLeader` command is proposed after proposing memory locks. And the raft transfer leader is triggered after the `TransferLeader` command is applied, which is after the memory-lock logs.
So, I don't think the unavailable duration should be related to how many lock requests there are... It seems to me that the unavailable duration due to rejecting proposals is about the same as now...

Reply:
> The TransferLeader command is proposed after proposing memory locks. And the raft transfer leader is triggered after the TransferLeader command is applied, which is after the memory-lock logs.
So, I don't think the unavailable duration should be related to how many lock requests there are... It seems to me that the unavailable duration due to rejecting proposals is about the same as now...

Agree.

> Is it possible that the leader fails to transfer leadership due to leader_transfer_max_log_lag even if you propose the transfer-leader request other than using raft message?

I think the possibility that rejects to transfer leader due to `leader_transfer_max_log_lag`  is the same as now with your PR, and it's wasteful to replicate memory locks in such a case.

> I'm thinking can we propose memory lock requests after the check and add some actions to reduce the unavailable duration due to it.

My point is how about proposing memory locks after passing the pre-transfer-leader check, if so, no waste but with longer unavailable duration, so we need some actions to reduce it...

Reply:
> I think the possibility that rejects to transfer leader due to leader_transfer_max_log_lag is the same as now with your PR, and it's wasteful to replicate memory locks in such a case.

Ah... I got what you mean... I was assuming the possibility of rejecting transfer leader is low, so it will unlikely to cause waste. But it seems not always the case after another thought...